### PR TITLE
api /addFiles

### PR DESCRIPTION
**Pull request by JayanthyChengan**
**What this PR does:**
A new API call /addFiles is added and it’s used to add metadata of multiple files to a given dataset in one call.

**why we need it:**
The existing API $SERVER_URL/api/datasets/:persistentId/add gives us limitation since it adds only one file to the dataset at a time.

**Which issue(s) this PR closes:**
Closes #7900

**Special notes for your reviewer:**

**Suggestions on how to test this:**

Please follow the steps in this documentation ( https://guides.dataverse.org/en/5.4.1/developers/s3-direct-upload-api.html ) to transfer multiple files to S3.  Once all the files exists in the s3 bucket, the new API /addFiles can be invoked by using the “jsonData” parameter

curl -H "X-Dataverse-key:xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx" -X POST  https://demo.dataverse.org/api/datasets/:persistentId/addFiles?persistentId=doi:10.5072/FK2/7U7YBV -F jsonData='[{"description":"My description.","directoryLabel":"data/subdir1","categories":["Data"], "restrict":"false", "storageIdentifier":"s3://demo-dataverse-bucket:176e28068b0-1c3f80357c42", "fileName":"file1.txt", "mimeType":"text/plain", "checksum": {"@type": "SHA-1", "@value": "123456"}}, {"description":"My description.","directoryLabel":"data/subdir1","categories":["Data"], "restrict":"false", "storageIdentifier":"s3://demo-dataverse-bucket:176e28068b0-1c3f80357d53", "fileName":"file2.txt", "mimeType":"text/plain", "checksum": {"@type": "SHA-1", "@value": "123789"}}]' 
 
**Does this PR introduce a user interface change? If mockups are available, please link/include them here:**  NO

**Is there a release notes update needed for this change?:** NO

**Additional documentation:**  added  notes to native API page

### CODE REVIEW

File: src/main/java/edu/harvard/iq/dataverse/ingest/IngestUtil.java [hunk 1]
```diff
@@ -257,7 +257,7 @@ public static String generateNewFileName(final String fileName) {
         // #6942 added proxy for existing files to a boolean set when dataset version copy is done
         for (Iterator<FileMetadata> fmIt = version.getFileMetadatas().iterator(); fmIt.hasNext();) {
             FileMetadata fm = fmIt.next();
-            if((fm.isInPriorVersion() || fm.getId() != null) && (replacedFmd==null) || (!fm.getDataFile().equals(replacedFmd.getDataFile()))) {
+            if((fm.isInPriorVersion() || fm.getId() != null) && (replacedFmd==null || !fm.getDataFile().equals(replacedFmd.getDataFile()))) {
```
Comment (qqmyers) [inline src/main/java/edu/harvard/iq/dataverse/ingest/IngestUtil.java:260]: Is this related to this PR?

Comment (JayanthyChengan) [reply src/main/java/edu/harvard/iq/dataverse/ingest/IngestUtil.java:260]: Implemented this change to solve the below exception. It got resolved,  but have to confirm with the team.


[2021-05-28T11:47:00.302-0400] [Payara 5.2021.1] [WARNING] [AS-EJB-00056] [javax.enterprise.ejb.container] [tid: _ThreadID=95 _ThreadName=http-thread-pool::http-listener-1(2)] [timeMillis: 1622216820302] [levelValue: 900] [[
  A system exception occurred during an invocation on EJB IngestServiceBean, method: public java.util.List edu.harvard.iq.dataverse.ingest.IngestServiceBean.saveAndAddFilesToDataset(edu.harvard.iq.dataverse.DatasetVersion,java.util.List,edu.harvard.iq.dataverse.DataFile)]]

javax.ejb.EJBException
	at com.sun.ejb.containers.EJBContainerTransactionManager.processSystemException(EJBContainerTransactionManager.java:723)
	at com.sun.ejb.containers.EJBContainerTransactionManager.completeNewTx(EJBContainerTransactionManager.java:652)
	at com.sun.ejb.containers.EJBContainerTransactionManager.postInvokeTx(EJBContainerTransactionManager.java:482)
	at com.sun.ejb.containers.BaseContainer.postInvokeTx(BaseContainer.java:4592)
	at com.sun.ejb.containers.BaseContainer.postInvoke(BaseContainer.java:2125)
	at com.sun.ejb.containers.BaseContainer.postInvoke(BaseContainer.java:2095)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandler.invoke(EJBLocalObjectInvocationHandler.java:220)
	at com.sun.ejb.containers.EJBLocalObjectInvocationHandlerDelegate.invoke(EJBLocalObjectInvocationHandlerDelegate.java:90)
	at com.sun.proxy.$Proxy352.saveAndAddFilesToDataset(Unknown Source)
	**at edu.harvard.iq.dataverse.ingest.__EJB31_Generated__IngestServiceBean__Intf____Bean__.saveAndAddFilesToDataset(Unknown Source)**
	at edu.harvard.iq.dataverse.datasetutility.AddReplaceFileHelper.step_060_addFilesViaIngestService(AddReplaceFileHelper.java:1563)
	at edu.harvard.iq.dataverse.datasetutility.AddReplaceFileHelper.runAddReplacePhase2(AddReplaceFileHelper.java:736)
	at edu.harvard.iq.dataverse.datasetutility.AddReplaceFileHelper.runAddReplaceFile(AddReplaceFileHelper.java:489)
	at edu.harvard.iq.dataverse.datasetutility.AddReplaceFileHelper.runAddFileByDataset(AddReplaceFileHelper.java:349)
	at edu.harvard.iq.dataverse.datasetutility.AddReplaceFileHelper.addFiles(AddReplaceFileHelper.java:2036)
	at edu.harvard.iq.dataverse.api.Datasets.addFilesToDataset(Datasets.java:2507)
	……………..
**Caused by: java.lang.NullPointerException
	at edu.harvard.iq.dataverse.ingest.IngestUtil.existingPathNamesAsSet(IngestUtil.java:260)**
	at edu.harvard.iq.dataverse.ingest.IngestUtil.checkForDuplicateFileNamesFinal(IngestUtil.java:66)
	at edu.harvard.iq.dataverse.ingest.IngestServiceBean.saveAndAddFilesToDataset(IngestServiceBean.java:167)

**Review by qqmyers — CHANGES_REQUESTED** [review]
This looks good to me. FWIW: I saw an earlier version of this and was able to create a branch of DVUploader (see https://github.com/GlobalDataverseCommunityConsortium/dataverse-uploader/tree/addFiles) that uses the new API to /addFiles for all files uploaded from a given folder in one call (versus calling /addFile for each one). 

I don't see anything but minor a minor style change needed (see  the reviewdog comment) but I'll note that the PR doesn't use the ~standard template - it would be good to have the 'Closes #' which will auto-close the issue when it is merged, and to have the other sections filled in as well (how to test, reviewer notes, is there a UI change, etc.) I assume you can cut/paste the template from another PR. (One note  - in the issue, the examples have "storageIdentifier2" in the json example which I think shouldn't have the 2. Having a good version of the json format in the PR will help @kcondon test).

**Review by pdurbin — COMMENTED** [review]
I haven't been following this pull request closely but we need an addition to the API Guide, right?

**Comment by JayanthyChengan** [comment]
@pdurbin :  Yes, I need to add new API call details in Native API guide.  I will work on it.  Thanks

**Review by pdurbin — COMMENTED** [review]
I haven't looked at the code yet but I left some comments on the docs. Thanks for adding them!

File: doc/sphinx-guides/source/api/native-api.rst [hunk 2]
```diff
@@ -2086,6 +2086,44 @@ The fully expanded example above (without environment variables) looks like this
...
+
+* "description" - A description of the file
+* "directoryLabel" - The "File Path" of the file, indicating which folder the file should be uploaded to within the dataset
+* "storageIdentifier" - String
+* "fileName" - String
+* "mimeType" - String
+* "fixity/checksum" either:
+
+  * "md5Hash" - String with MD5 hash value, or
+  * "checksum" - Json Object with "@type" field specifying the algorithm used and "@value" field with the value from that algorithm, both Strings
```
[the remark this answers is not in the document]
Comment (JayanthyChengan) [reply doc/sphinx-guides/source/api/native-api.rst:2106]: Hi @pdurbin : The jsonString format in this new /addFiles API code follows same structure as in $SERVER_URL/api/datasets/:persistentId/add API.  

So, for the jsonString, I referred to the doc:
https://guides.dataverse.org/en/5.5/developers/s3-direct-upload-api.html under the section "Adding the Uploaded file to the Dataset" .

[the remark this answers is not in the document]
Comment (pdurbin) [reply doc/sphinx-guides/source/api/native-api.rst:2106]: Ah, so the @ signs originally came from @qqmyers I would assume. I don't have anything against @ signs. 😄 I was just wondering why they're there.

File: doc/sphinx-guides/source/api/native-api.rst [hunk 3]
```diff
@@ -2086,6 +2086,44 @@ The fully expanded example above (without environment variables) looks like this
...
+This API call requires a ``jsonString`` expressing the metadata of multiple files.  It adds file metadata to the database table where the file already copied to the storage.
+
+The jsonData object includes values for:
+
+* "description" - A description of the file
+* "directoryLabel" - The "File Path" of the file, indicating which folder the file should be uploaded to within the dataset
+* "storageIdentifier" - String
+* "fileName" - String
+* "mimeType" - String
+* "fixity/checksum" either:
```
[the remark this answers is not in the document]
Comment (JayanthyChengan) [reply doc/sphinx-guides/source/api/native-api.rst:2133]: @pdurbin - The jsonData in this api call follows the same structure as mentioned in this guide.
https://guides.dataverse.org/en/latest/developers/s3-direct-upload-api.html  under "Adding the Uploaded file to the Dataset" section

And @qqmyers implemented the support for different checksum at this PR https://github.com/IQSS/dataverse/pull/7602

[the remark this answers is not in the document]
Comment (qqmyers) [reply doc/sphinx-guides/source/api/native-api.rst:2133]: To be clear, the choice isn't fixity vs. checksum, it's whether there's an md5Hash element with a string value (the hash) (legacy/backward compatible) or a checksum element with an object value where the algorithm (@type) and hash (@value) are specified.

File: doc/sphinx-guides/source/api/native-api.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (qqmyers) [reply doc/sphinx-guides/source/api/native-api.rst:2136]: @type and @value are reserved words in json-ld,  and are how the checksum object is represented in the OAI-ORE export, but they are also 'just json' and don't involve any json-ld generation/parsing is required here.

[the remark this answers is not in the document]
Comment (JayanthyChengan) [reply doc/sphinx-guides/source/api/native-api.rst:2136]: Thanks @qqmyers

File: doc/sphinx-guides/source/api/native-api.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (pdurbin) [reply doc/sphinx-guides/source/api/native-api.rst:2133]: Thanks to both. These docs aren't super clear, in my opinion. Preexisting condition. 😄

File: doc/sphinx-guides/source/api/native-api.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (pdurbin) [reply doc/sphinx-guides/source/api/native-api.rst:2136]: Ok, to me it's weird to have a JSON document (below) that's 90% vanilla JSON and 10% "looks like JSON-LD with @ signs". But this is preexisting. Not worth opening that can of worms. 😄 

`"{'description':'My description.','directoryLabel':'data/subdir1','categories':['Data'], 'restrict':'false', 'storageIdentifier':'s3://demo-dataverse-bucket:176e28068b0-1c3f80357c42', 'fileName':'file1.txt', 'mimeType':'text/plain', 'checksum': {'@type': 'SHA-1', '@value': '123456'}}"`

File: src/main/java/edu/harvard/iq/dataverse/datasetutility/AddReplaceFileHelper.java [hunk 4]
```diff
@@ -5,16 +5,11 @@
  */
 package edu.harvard.iq.dataverse.datasetutility;
 
-import edu.harvard.iq.dataverse.DataFile;
+import edu.harvard.iq.dataverse.*;
```
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/datasetutility/AddReplaceFileHelper.java:8]: We generally avoid wildcard imports. If you feel like changing it back, great. Otherwise, no big deal.

File: src/main/java/edu/harvard/iq/dataverse/datasetutility/AddReplaceFileHelper.java [hunk 5]
```diff
@@ -1959,6 +1989,160 @@ public String getDuplicateFileWarning() {
...
+                DatasetLock eipLock = dataset.getLockFor(DatasetLock.Reason.EditInProgress);
+                if (eipLock == null) {
+                    logger.log(Level.WARNING, "Dataset not locked for EditInProgress ");
+                } else {
+                    datasetService.removeDatasetLocks(dataset, DatasetLock.Reason.EditInProgress);
+                    logger.log(Level.INFO, "Removed EditInProgress lock ");
+                }
+
+                try {
+                    Command<Dataset> cmd = new UpdateDatasetVersionCommand(dataset, dvRequest);
```
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/datasetutility/AddReplaceFileHelper.java:2103]: It's not super clear from the existing docs what permissions are required (at least from a quick reading just now) so I'm glad to see that UpdateDatasetVersionCommand is required. I was starting to worry that it's wide open. If you feel like adding something to the docs about this, please go ahead.

File: src/main/java/edu/harvard/iq/dataverse/ingest/IngestUtil.java [hunk 6]
```diff
@@ -57,7 +57,7 @@
      *
      * @param version the dataset version
      * @param newFiles the list of new data files to add to it
-     * @param fileToReplace 
+     * @param fileToReplace
```
Comment (pdurbin) [inline src/main/java/edu/harvard/iq/dataverse/ingest/IngestUtil.java:60]: I'm just putting this comment here near the end of the pull request because it's where I look for tests. If it's easy to add an integration test, please go ahead. I can see how this could be tricky because you've got to get the file to add onto the disk of the remote server somehow. So please don't worry about it if it's difficult. This could be worked on separately for all this related code (single file add, etc.),

**Review by pdurbin — CHANGES_REQUESTED** [review]
@JayanthyChengan I should have asked for this earlier but can you please add a release note?

I also left you some other comments but you should please consider them all optional.

I'm excited about getting closer to Globus support!

File: src/main/java/edu/harvard/iq/dataverse/datasetutility/AddReplaceFileHelper.java (hunk 4 above)
Comment (JayanthyChengan) [reply src/main/java/edu/harvard/iq/dataverse/datasetutility/AddReplaceFileHelper.java]: @pdurbin - I will submit the corrections.

File: src/main/java/edu/harvard/iq/dataverse/datasetutility/AddReplaceFileHelper.java (hunk 5 above)
Comment (JayanthyChengan) [reply src/main/java/edu/harvard/iq/dataverse/datasetutility/AddReplaceFileHelper.java:2116]: Updated native-api.rst.  Please review

File: src/main/java/edu/harvard/iq/dataverse/ingest/IngestUtil.java (hunk 6 above)
Comment (JayanthyChengan) [reply src/main/java/edu/harvard/iq/dataverse/ingest/IngestUtil.java:60]: Sorry, this is not implemented in this PR.

File: doc/sphinx-guides/source/api/native-api.rst [hunk 7]
```diff
@@ -2116,6 +2116,44 @@ The fully expanded example above (without environment variables) looks like this
 
 Note: The ``id`` returned in the json response is the id of the file metadata version.
 
+
```
[the remark this answers is not in the document]
Comment (JayanthyChengan) [reply doc/sphinx-guides/source/api/native-api.rst:2119]: added doc/release-notes/7900-add-multipleFilesMetadata-dataset.md

**Review by pdurbin — APPROVED** [review]
I haven't run the code or built the docs myself but this seems ready for QA now. Thanks, @JayanthyChengan !

File: src/main/java/edu/harvard/iq/dataverse/ingest/IngestUtil.java (hunk 6 above)
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/ingest/IngestUtil.java:60]: No worries.

File: src/main/java/edu/harvard/iq/dataverse/datasetutility/AddReplaceFileHelper.java (hunk 5 above)
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/datasetutility/AddReplaceFileHelper.java:2116]: Looks good. Thanks.
